packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD2.hpp

```diff
@@ -0,0 +1,426 @@
+            size_t vid_d2_degree = 0;
+	    //calculate the degree for degree-based recoloring
+            if(vid_d2 < n_local){
+              vid_d2_degree = dist_offsets(vid_d2+1) - dist_offsets(vid_d2);
+            } else {
+              vid_d2_degree = ghost_degrees(vid_d2-n_local);
+            }
+            if(curr_lid != vid_d2 && femv_colors(vid_d2) == curr_color){
+              if(curr_degree < vid_d2_degree && recolor_degrees){
+                found = true;
+                femv_colors(curr_lid) = 0;
+                recoloring_size++;
+                break;//---------------------------------------------------
+              } else if(vid_d2_degree < curr_degree && recolor_degrees){//|
+                femv_colors(vid_d2) = 0;                                //|
+                recoloring_size++;                                      //|
+              } else if(rand(curr_lid) < rand(vid_d2)){                 //|
+                found = true;                                           //|
+                femv_colors(curr_lid) = 0;                              //|
+                recoloring_size++;                                      //|
+                break;//--------------------------------------------------|
+              } else if(rand(vid_d2) < rand(curr_lid)){                 //|
+                femv_colors(vid_d2) = 0;                                //|
+                recoloring_size++;                                      //|
+              } else {                                                  //|
+                if(gid(curr_lid) >= gid(vid_d2)){                       //|
+                  found = true;                                         //|
+                  femv_colors(curr_lid) = 0;                            //|
+                  recoloring_size++;                                    //|
+                  break;//------------------------------------------------|
+                } else {                                                //|
+                  femv_colors(vid_d2) = 0;                              //|
+                  recoloring_size++;//                                    v
+                }//              If we uncolor the vertex whose neighbors we're
+              }  //              checking, each subsequent conflict check will 
+            }    //              not do anything productive. We need this------
+          }      //              to completely move on to the next vertex.    |
+          if(found) break;//<--------------------------------------------------
+        }
+      },recoloringSize(0));
+      Kokkos::fence();
+
+      //update the verts_to_send and verts_to_recolor views.
+      Kokkos::parallel_for(femv_colors.size(), KOKKOS_LAMBDA(const uint64_t& i){
+        if(femv_colors(i) == 0){
+	  //we only send vertices owned by the current process
+          if(i < n_local){
+            verts_to_send_atomic(verts_to_send_size_atomic(0)++) = i;
```

Comment:
POSSIBLE BUG:  Are we sure that this syntax will not have race conditions?  Could the following occur:

1. Thread 1 gets verts_to_send_size_atomic(0)
2. Thread 2 gets verts_to_send_size_atomic(0)
3. Both threads atomically change the same entry of verts_to_send_atomic (one after the other)
4. Both threads atomically increment verts_to_send_size_atomic (one after the other)

I would have expected something like the following

```
auto idx = atomic_fetch_add(verts_to_send_size_atomic(0), 1);
verts_to_send_atomic(idx) = i;
```
In either case, why does verts_to_send_atomic need to be atomic?  Each entry should be written only once, right?  Maybe @gslota can provide some insight here.

Reply:
@crtrott confirmed that that using view_atomic(0)++ is equivalent to atomic_fetch_add.  So there is no bug and no need to change to atomic_fetch_add.  But verts_to_send_atomic does not need to be atomic, right?  We write each entry only once.  Similar with verts_to_recv_atomic.

Reply:
Correct. This change has been made across all files, and the atomic versions of the array Kokkos views have been removed from the code. Autotesting shows that there are likely no race conditions in the new code.